registry/session_v2.go

```diff
@@ -95,11 +98,43 @@ func (r *Session) GetV2ImageManifest(ep *Endpoint, imageName, tagName string, au
 		return nil, "", utils.NewHTTPRequestError(fmt.Sprintf("Server error: %d trying to fetch for %s:%s", res.StatusCode, imageName, tagName), res)
 	}
 
-	buf, err := ioutil.ReadAll(res.Body)
+	manifestBytes, err := ioutil.ReadAll(res.Body)
 	if err != nil {
 		return nil, "", fmt.Errorf("Error while reading the http response: %s", err)
 	}
-	return buf, res.Header.Get(DockerDigestHeader), nil
+
+	sig, err := libtrust.ParsePrettySignature(manifestBytes, "signatures")
+	if err != nil {
+		return nil, "", fmt.Errorf("error parsing signed manifest: %s", err)
+	}
+
+	payload, err := sig.Payload()
+	if err != nil {
+		return nil, "", fmt.Errorf("error retrieving signed manifest payload: %s", err)
+	}
+
+	hdrDigest, err := digest.ParseDigest(res.Header.Get(DockerDigestHeader))
+	if err != nil {
+		return nil, "", fmt.Errorf("invalid manifest digest from registry: %s", err)
+	}
+
+	dgstVerifier, err := digest.NewDigestVerifier(hdrDigest)
+	if err != nil {
+		return nil, "", fmt.Errorf("unable to verify manifest digest from registry: %s", err)
+	}
+
+	dgstVerifier.Write(payload)
+
+	if !dgstVerifier.Verified() {
+		computedDigest, _ := digest.FromBytes(payload)
```

Comment:
Is the payload guaranteed to be the same as what the registry is storing?
Any chance the fields could show up in a different order?
On Wed, Mar 18, 2015 at 6:05 PM Josh Hawn <notifications@github.com> wrote:

> In registry/session_v2.go
> <https://github.com/docker/docker/pull/11457#discussion_r26712582>:
>
> > +	}
> > +
> > +	hdrDigest, err := digest.ParseDigest(res.Header.Get(DockerDigestHeader))
> > +	if err != nil {
> > +		return nil, "", fmt.Errorf("invalid manifest digest from registry: %s", err)
> > +	}
> > +
> > +	dgstVerifier, err := digest.NewDigestVerifier(hdrDigest)
> > +	if err != nil {
> > +		return nil, "", fmt.Errorf("unable to verify manifest digest from registry: %s", err)
> > +	}
> > +
> > +	dgstVerifier.Write(payload)
> > +
> > +	if !dgstVerifier.Verified() {
> > +		computedDigest, _ := digest.FromBytes(payload)
>
> If the registry has gone rogue, then maybe.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/11457/files#r26712582>.
>